File: Python/lld-examples/medium/restaurant-management-system/table.py
from typing import Optional
from enum import Enum

class TableStatus(Enum):
    """Table status enum"""
    AVAILABLE = "Available"
    OCCUPIED = "Occupied"
    RESERVED = "Reserved"
    OUT_OF_SERVICE = "Out of Service"

class Table:
    """Table class representing a restaurant table"""
    
    def __init__(self, table_number: int, capacity: int):
        self.table_number = table_number
        self.capacity = capacity
        self.status = TableStatus.AVAILABLE
        self.current_customer = None
        self.current_order = None
        self.reservation_time = None
    
    def is_available(self) -> bool:
        """Check if table is available"""
        return self.status == TableStatus.AVAILABLE
    
    def reserve_table(self, customer):
        """Reserve table for customer"""
        if self.is_available():
            self.status = TableStatus.OCCUPIED
            self.current_customer = customer
            return True
        return False
    
    def clear_table(self):
        """Clear table after customer leaves"""
        self.status = TableStatus.AVAILABLE
        self.current_customer = None
        self.current_order = None
        self.reservation_time = None
    
    def set_out_of_service(self):
        """Set table out of service"""
        self.status = TableStatus.OUT_OF_SERVICE
    
    def set_available(self):
        """Set table available"""
        self.status = TableStatus.AVAILABLE
    
    def __str__(self):
        return f"Table {self.table_number} - {self.capacity} seats - {self.status.value}" 

File: JAVA/lld-examples/medium/restaurant-management-system/src/main/java/com/systemdesign/lld/restaurant/Table.java
package com.systemdesign.lld.restaurant;

import java.time.LocalDateTime;

/**
 * Represents a table in the restaurant
 */
public class Table {
    private int tableNumber;
    private int capacity;
    private TableStatus status;
    private LocalDateTime reservationTime;
    private Customer customer;
    private Order currentOrder;
    
    public Table(int tableNumber, int capacity) {
        this.tableNumber = tableNumber;
        this.capacity = capacity;
        this.status = TableStatus.AVAILABLE;
        this.reservationTime = null;
        this.customer = null;
        this.currentOrder = null;
    }
    
    public boolean isAvailable() {
        return status == TableStatus.AVAILABLE;
    }
    
    public boolean canAccommodate(int partySize) {
        return partySize <= capacity;
    }
    
    public void reserve(Customer customer, LocalDateTime reservationTime) {
        this.customer = customer;
        this.reservationTime = reservationTime;
        this.status = TableStatus.RESERVED;
    }
    
    public void occupy() {
        this.status = TableStatus.OCCUPIED;
    }
    
    public void free() {
        this.status = TableStatus.AVAILABLE;
        this.customer = null;
        this.reservationTime = null;
        this.currentOrder = null;
    }
    
    // Getters and setters
    public int getTableNumber() { return tableNumber; }
    public int getCapacity() { return capacity; }
    public TableStatus getStatus() { return status; }
    public LocalDateTime getReservationTime() { return reservationTime; }
    public Customer getCustomer() { return customer; }
    public Order getCurrentOrder() { return currentOrder; }
    public void setCurrentOrder(Order order) { this.currentOrder = order; }
    
    @Override
    public String toString() {
        return String.format("Table %d (Capacity: %d, Status: %s)", 
            tableNumber, capacity, status);
    }
}

enum TableStatus {
    AVAILABLE,
    RESERVED,
    OCCUPIED
} 